### PR TITLE
feat(datasets) Make complex_column_expr parametric to reuse it for the AST

### PR DESCRIPTION
**Pull request by fpacifici**
complex_column_expr parses nested expressions in the current system. In the current system parsing and formatting happen together.
I think we should preserve the parsing part of it (that is dependent on the syntax) and change the outcome of that function. 

This PR makes complex_column_expr parametric, it now takes three functions as parameters that are used to produce the output. It also adds a rudimentary implementation to generate Expressions to prove the concept.

Test plan:
complex_column_expr is well covered. The tests still pass. A clone of that test is added for expressions.

### CODE REVIEW

File: snuba/query/parser/functions.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            elif isinstance(nxt, str):
+                out.append(simple_expression_builder(nxt))
+            else:
+                out.append(literal_builder(nxt))
+            i += 1
+
+    return output_builder(alias, name, out)
+
+
+def parse_function_to_expr(expr: Any) -> Expression:
```
Comment (fpacifici) [inline snuba/query/parser/functions.py]: This method is here only for the sake of being able to test parsing. The real parser will look different.
See the test to have an idea of how this works.

File: snuba/query/parser/functions.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+from snuba.util import is_function, QUOTED_LITERAL_RE
+
+TExpression = TypeVar("TExpression")
+
+
+def parse_function(
+    output_builder: Callable[[Optional[str], str, List[TExpression]], TExpression],
+    simple_expression_builder: Callable[[str], TExpression],
+    literal_builder: Callable[
+        [Optional[Union[str, datetime, date, List[Any], Tuple[Any], numbers.Number]]],
```
Comment (tkaemming) [inline snuba/query/parser/functions.py]: extreme nit: I don't think we are planning on supporting all types of number (e.g. `Complex`.)

**Review by tkaemming — APPROVED** [review]
This makes sense and is a clever way to approach this.

File: snuba/query/parser/functions.py (hunk 2 above)
Comment (fpacifici) [reply snuba/query/parser/functions.py]: Indeeed, unfortunately that signature is required by this method 
https://github.com/getsentry/snuba/blob/master/snuba/util.py#L202-L222

which is called by that function

**Pull request merged**